docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst

```diff
@@ -0,0 +1,1946 @@
+except that they also contain structured data within them.
+
+For example, the following variant has two constructors, each with
+unique fields. Overwrite both v1 and v2 modules with the following
+source:
+
+.. code:: daml
+
+  module Main where
+  
+  data Shape
+    = Circle
+    | Polygon { sides : Int }
+
+We can extend this variant in two ways. We can add a new constructor,
+similarly to enums. Modify the v2 module to add a new ``Line`` constructor
+with a ``len`` field:
+
+.. code:: daml
+
+  module Main where
+  
+  data Shape
+    = Circle
+    | Polygon { sides : Int }
+    | Line { len : Numeric 10 } -- New line constructor
+
+As before, building should succeed.
+
+.. code:: bash
+
+  > cd v1/my-pkg
+  > daml build
+  ...
+  Created .daml/dist/my-pkg-1.0.0.dar
+  > cd ../../v2/my-pkg
+  > daml build
+  ...
+  Created .daml/dist/my-pkg-2.0.0.dar
+
+We can also add a new field to a constructor, similarly to templates -
+for example, add a ``sideLen`` field to our ``Polygon`` constructor, to specify
+the lengths of the sides of the polygon.
+
+.. code:: daml
+
+  data Shape
+    = Circle
+    | Polygon
+        { sides : Int
+        , sideLen : Numeric 10 -- New field
+        }
+    | Line { len : Numeric 10 }
+
+Building this fails because the new ``sideLen`` field is non-optional.
+
+.. code:: bash
+
+  > cd v2/my-pkg
+  > daml build
+  ...
+  error type checking data type Main.Shape:
+    The upgraded variant constructor Polygon from variant Shape has added a field.
+  ERROR: Creation of DAR file failed.
+
+Making the new ``sideLen`` field optional fixes the error:
+
+.. code:: daml
+
+  ...
+        , sideLen : Optional (Numeric 10) -- New field
+  ...
+
+.. code:: bash
+
+  > cd v2/my-pkg
+  > daml build
+  ...
+  Created .daml/dist/my-pkg-2.0.0.dar
+
+Limitations in Upgrading Variants
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Upgrading variants has some limitations - because we have defined the
+``Circle`` constructor without a field in curly braces, it cannot be
+upgraded with new fields.
+
+.. code:: daml
+
+  ...
+    -- Add a field where no fields existed
+    = Circle { radius : Optional (Numeric 10) }
+  ...
+
+.. code:: bash
+
+  > cd v2/my-pkg
+  > daml build
+  ...
+  error type checking data type Main.Shape:
+    The upgraded data type Shape has changed the type of a variant.
+  ERROR: Creation of DAR file failed.
+
+The same applies to variants with unnamed fields. If the v1 definition
+of the ``Line`` constructor were as follows, it would also not be able to
+upgrade:
+
+.. code:: daml
+
+  ...
+    | Line (Numeric 10)
+
+In general, in order to enable future upgrades, it is strongly
+recommended that all constructors use named fields, and that all
+constructors have at least one field. If a constructor has no fields in
+an initial v1 package, one can assign a dummy field.
+
+For example, the correct way to write the v1 ``Circle`` constructor would be
+as follows:
+
+.. code:: daml
+
+  ...
+    = Circle { dummy : () }
```

Comment:
@dylant-da

Reply:
The compiler doesn't add a default dummy field for variants - it instead pretends you set the variant's inner type to Unit, i.e. writing `= Circle` and `= Circle ()` are identical in terms of what the engine eventually processes.

We planned to allow Unit to be upgradeable to anything, but deferred it due to time constraints and because it is a reverse-compatible change so we can implement it in 2.10 without breaking anything in 2.9.

Reply:
Do we consider this an explicit limitation, or simply an additional feature that isn't in 2.9

Reply:
It is a limitation, but not one that should go into notes - it is already clarified as a limitation in the section on variant upgrades.

Reply:
Right, cheers